mtags/src/main/scala/scala/meta/internal/IdentifierComparator.scala

```diff
@@ -14,35 +16,47 @@ import java.util.Comparator
 object IdentifierComparator extends Comparator[CharSequence] {
   override def compare(o1: CharSequence, o2: CharSequence): Int = {
     val len = math.min(o1.length(), o2.length())
-    var i = 0
-    while (i < len) {
-      val a = o1.charAt(i)
-      val b = o2.charAt(i)
-      if (a.isDigit && b.isDigit) {
-        val byDigit = Integer.compare(toDigit(o1, i), toDigit(o2, i))
-        if (byDigit != 0) return byDigit
-        else {
-          i = seekNonDigit(o1, i)
-        }
-      } else {
-        val result = Character.compare(a, b)
-        if (result != 0) {
-          return result
+
+    @tailrec
+    def compareLoop(idx: Int): Int = {
+      if (idx >= len) Integer.compare(o1.length(), o2.length())
+      else {
+        val a = o1.charAt(idx)
+        val b = o2.charAt(idx)
+        if (a.isDigit && b.isDigit) {
+          val byDigits = compareSequences(o1, o2, idx)
+          if (byDigits != 0) byDigits
+          else compareLoop(seekNonDigit(o1, idx))
+        } else {
+          val result = Character.compare(a, b)
+          if (result != 0) result
+          else compareLoop(idx + 1)
         }
-        i += 1
       }
     }
-    Integer.compare(o1.length(), o2.length())
+
+    compareLoop(0)
   }
-  private def seekNonDigit(cs: CharSequence, i: Int): Int = {
-    var curr = i
-    while (curr < cs.length() && cs.charAt(curr).isDigit) {
-      curr += 1
-    }
-    curr
+
+  @tailrec
+  def seekNonDigit(cs: CharSequence, idx: Int): Int = {
+    val condition = idx < cs.length() && cs.charAt(idx).isDigit
+    if (condition) seekNonDigit(cs, idx + 1)
+    else idx
+  }
+
+  private def compareSequences(
+      s1: CharSequence,
+      s2: CharSequence,
+      idx: Int
+  ): Int = {
+    val first = toDigit(s1, idx)
+    val second = toDigit(s2, idx)
+    first.compare(second)
   }
-  private def toDigit(cs: CharSequence, i: Int): Int = {
+
+  private def toDigit(cs: CharSequence, i: Int): BigInt = {
     val digit = cs.subSequence(i, seekNonDigit(cs, i))
-    Integer.parseUnsignedInt(digit.toString)
+    BigInt(digit.toString())
```

Comment:
Will BigInt not cause a bit of performance regression? This is used in completions, so I would love to avoid it.

There was some discusssion on the topic here: https://contributors.scala-lang.org/t/optimizing-bigint-for-small-integers/3412/3

However it seems this would only be available for Scala 2.13.

What if instead, we first compared the lengths from `seekNonDigit(cs, i) - i`and then if it's the same compare digits until one is bigger?

Reply:
>  This is used in completions

Ahhh, I forgot about that. I was worried about performance but then I thought: it's just TVP, it's not an issue.

I was thinking about a similar approach, I'll do that.